### PR TITLE
Sync to cloud: raw data, docs, articles, embeddings, search index

### PR DESCRIPTION
**Pull request by jwayne**
I updated our `preprocesor.py`, `benchmark_*.py`, and Django code so all data is loaded from GCS or Google Firestore (NoSQL db). This has 2 major benefits:
* Much easier to collaborate, without needing to manually upload or download data files
* Docker image no longer contains any data, so building it is much faster. Rather, image pulls data from GCS.

Note that our GCS integration uses a local file cache (stored at `backend/data/gcs_cache`), so the same machine won't download the same file multiple times.

Here's how I treat each type of data:
* raw data: stored in GCS at `gs://streamlitdata/raw`
    * **Change in process:** When we scrape new data, we should upload raw JSON files to GCS rather than GDrive. This is easy to do via the following link: https://console.cloud.google.com/storage/browser/streamlitdata/raw
    * `preprocessor.py` now loads raw files from GCS.
* embeddings: each chunk's embedding is individually stored in Google Firestore (NoSQL database), viewable [here](https://console.cloud.google.com/firestore/databases/-default-/data/panel/embedding/00525b4b2be0d29fac47ed16d1e555a8214e69202d5b6e4f585f03a62936031c?authuser=2&project=streamlitlogs). The ID of each Firestore entry matches the chunk text's SHA-256 hash, so it's easy to identify when a chunk's embedding needs to be recomputed. The `embedding` value is stored as a pickle string. I chose Firestore because batch reads and writes are very fast.
* docs, articles: stored in GCS at `gs://streamlitdata`
    * `preprocessor.py` writes computed docs/chunks to GCS, or alternatively skips loading raw files or computing chunks if computed docs/chunks already exist.
    * `data_loader.py` now always loads these files from GCS (for use in `benchmark_*.py` scripts + Django app)
*  search index: stored in GCS at `gs://streamlitdata`, similar to docs+articles

As part of this, I made some major changes to the following code:
* `SemanticSearch` -> `SearchFactory.get_search_client_cls`. I think this helps separate the build/save/load logic for different NN algos into different subclasses, rather than all lumped into `SemanticSearch` (in the case of build/load) or split into `preprocessor.py` (in the case of save).
    * Instead of using `SemanticSearch(...)` to create an instance, we should instead use `SearchFactory`. See example use in `benchmark_*.py`
* `data_loader.py`: I deleted the file handlers since they're not longer needed.

### CODE REVIEW

**Comment by RomanEngeler1805** [comment]
Nice job!

I have seen that you moved the file handling to `core/utils` and the `core/file_handlers` is empty by now. I guess we can delete it

**Comment by jwayne** [comment]
Yes, exactly. Git won't always delete the folder - perhaps you have some cache files in there

File: .dockerignore [hunk 1]
```diff
@@ -17,11 +17,11 @@ db.sqlite3
 **/static
 
 # Project files to ignore
+/backend/data
 /backend/data/raw
 /backend/data/processed_old
 /backend/data/embeddings
-/backend/crawler
```
[the remark this answers is not in the document]
Comment (jwayne) [reply .dockerignore:23]: files only contain Python code, which makes a negligible impact to include. I think it's easier if the Docker image just contains all Python code in the repo.

File: backend/core/utils/retry.py
[the remark this answers is not in the document]
Comment (jwayne) [reply backend/core/utils/retry.py:1]: I thought Tenacity is more of a retry library. Or is it also logging?

File: backend/core/search/nn/annoy_client.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            data = pickle.load(f)
+
+        index = AnnoyIndex(embed_client.embedding_dim, "angular")
+        index.load(data)
+
+        return cls(index=index, embed_client=embed_client, k_pages=k_pages)
+
+    def save(self, file_name: str, gcs_bucket_name: str):
+        raise NotImplementedError
+        # TODO save annoy to pickle
```
[the remark this answers is not in the document]
Comment (jwayne) [reply backend/core/search/nn/annoy_client.py:57]: Great, that's what I figured!

File: backend/core/search/embed/sentence_bert.py [hunk 3]
```diff
@@ -29,8 +28,3 @@ def embed_query(self, text: str) -> List[float]:
         """Embed query text."""
         embeddings = self.client.encode(text)
         return embeddings
-
-    def parallel_embed_queries(self, texts: List[str]) -> List[List[float]]:
```
[the remark this answers is not in the document]
Comment (jwayne) [reply backend/core/search/embed/sentence_bert.py:33]: This code was copy-pasted in `OpenAIEmbeddings` and `CohereEmbeddings`. To avoid code duplication, we should rather implement it a single time in the base class. I was intending to do this, but once I implemented the new openai and cohere client classes, I realized that async made more sense than threading (the majority of time spent is in making Web requests, and the remaining code is forced to be single-thread anyway due to GIL). So I instead implemented  `async_embed_documents` in the openai and cohere clients, and never implemented the parallel method in the base class (since it wouldn't make sense for openai/cohere).

Later, I also examined how `parallel_embed_queries` was called, and noticed that it was only being called to embed **documents** (in `preprocessor.py`). While this is OK since we're currently using symmetric embedding, this inconsistency + the fact that I changed `preprocessor.py` to call the async method rather than the parallelization method + the fact we're not using SBert right now led me to just decide to remove this entirely.

If we do go back to using an open-source embedding model, I agree parallel threading is useful. That said, I'm not sure this logic should be in the SBert subclass - it might rather make sense for the caller to implement parallelization - since the caller may want to control the max number of threads, etc.

TLDR: for now, I suggest we remove this, and if we go back to open-source embedding, we can implement threading then and decide where to write that code

File: backend/core/search/embed/openai_embeddings.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    Computes embeddings using OpenAI's API.
+    """
+
+    embed_client_name = "openai"
+
+    def __init__(self, model_name=MODEL_NAME):
+        self.model_name = model_name
+        self.embedding_dim = EMBEDDING_DIMS[model_name]
+        self.max_seq_length = 8192  # Not sure what this is used for...
```
[the remark this answers is not in the document]
Comment (jwayne) [reply backend/core/search/embed/openai_embeddings.py:28]: Cool, i added a comment: `Currently not used, but could be used to manage max chunk size for context`

File: backend/core/utils/retry.py
[the remark this answers is not in the document]
Comment (RomanEngeler1805) [reply backend/core/utils/retry.py:1]: yeah should be retry
